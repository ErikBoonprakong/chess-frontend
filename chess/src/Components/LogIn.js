import React from "react";
import Networking from "./Networking.js";
import { Link, Redirect } from "react-router-dom";
import cookieObj from "./GetCookies";
import Alert from "react-bootstrap/Alert";
import "./Login.css";

class LogIn extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      username: "",
      password: "",
      error: "",
      redirect: false,
      valid: true,
      cookies: cookieObj(),
    };
    this.Networking = new Networking();
  }

  handleChange = async (event) => {
    await this.setState({ [event.target.id]: event.target.value });
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    const response = await this.Networking.postLogin(
      this.state.username,
      this.state.password
    );

    this.setState({ cookies: cookieObj() });

    this.props.newCookie(this.state.cookies);
    if (this.state.cookies.user) {
      this.setState({ redirect: true, valid: true });
    } else {
      this.setState({ valid: false, error: response.message });
    }
  };

  getWarning(input) {
    return <Alert variant="danger">{this.state.error}</Alert>;
  }

  render() {
    return (
      <div className="log-form">
        {this.state.redirect ? (
          <Redirect to="/home" />
        ) : (
          <div className="login-form">
            <h2>Login</h2>{" "}
            <form onSubmit={this.handleSubmit} data-testid="form">
              <div id="inputWithIcon">
                <input
                  name="username"
                  type="text"
                  placeholder="Username"
                  className="user"
                  id="username"
                  value={this.state.username}
                  onChange={this.handleChange}
                />
              </div>
              <div id="inputWithIcon">
                <input
                  name="password"
                  data-testid="password"
                  placeholder="Password"
                  type={this.state.showPassword ? "text" : "password"}
                  className="user"
                  id="password"
                  value={this.state.password}
                  onChange={this.handleChange}
                />
              </div>
              <br />
              <br />

              <div className="buttons">
                <input
                  className="login-btn-register-btn"
                  type="submit"
                  value="Login"
                ></input>
                <Link to="/register">
                  <button data-testid="register-btn" className="register-btn">
                    Create Account
                  </button>
                </Link>
                <h4 className="error-message">
                  {" "}
                  {!this.state.valid && this.getWarning()}
                </h4>
              </div>
            </form>
          </div>
        )}
      </div>
    );
  }
}

export default LogIn;
