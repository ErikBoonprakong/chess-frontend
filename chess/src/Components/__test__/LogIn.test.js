import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import {
  render,
  screen,
  fireEvent,
  waitForElement,
} from "@testing-library/react";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { configure } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import LogIn from "../LogIn";
import cookieObj from "../GetCookies";

configure({ adapter: new Adapter() });

describe("Login page form should include all elements", () => {
  test("Username form element must load", () => {
    render(
      <Router>
        <LogIn />
      </Router>
    );
    const input = screen.getByPlaceholderText("Username");
    expect(input).toBeInTheDocument();
  });
});
test("Password form element must load", () => {
  render(
    <Router>
      <LogIn />
    </Router>
  );
  const password = screen.getByPlaceholderText("Password");
  expect(password).toBeInTheDocument();
});

test("Register button must load", () => {
  render(
    <Router>
      <LogIn />
    </Router>
  );
  const button = screen.getByTestId("register-btn");
  expect(button).toBeInTheDocument();
});

describe("Log In", () => {
  let wrapper;
  let sessionCookie;
  function getCookie(newCookie) {
    sessionCookie = newCookie;
  }
  beforeEach(() => {
    wrapper = shallow(
      <LogIn newCookie={(newCookie) => getCookie(newCookie)} />
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("handleChangeTest", () => {
    it("should call setState on username", () => {
      const mockUsername = {
        target: {
          id: "username",
          value: "test",
        },
      };
      const mockPassword = {
        target: {
          id: "password",
          value: "testtest",
        },
      };
      const expected = {
        username: "test",
        password: "testtest",
      };
      wrapper.instance().handleChange(mockUsername);
      wrapper.instance().handleChange(mockPassword);
      expect(wrapper.state().username).toEqual(expected.username);
      expect(wrapper.state().password).toEqual(expected.password);
    });
  });
});
