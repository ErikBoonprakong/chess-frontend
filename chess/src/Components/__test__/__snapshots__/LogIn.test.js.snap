// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log In should match the snapshot 1`] = `ShallowWrapper {}`;
